File: problems/876_middle_of_the_linked_list.py
class Solution:
    def middleNode(self, head):
        list_len = 0
        old_head = head
        while head:
            list_len += 1
            head = head.next
            middle_index = list_len // 2
        while middle_index != 0:
            middle_index -= 1
            old_head = old_head.next
        return old_head

File: leetcode/java/876MiddleOfTheLinkedList.java
/**
 * T: O(N) S: O(1)
 *
 * <p>Move two pointers, one at the normal speed, another one at the double speed. When the latter
 * reaches the end of the list - the former one points to the middle.
 */
public class ListNode {
  int val;
  ListNode next;

  ListNode(int x) {
    val = x;
  }
}

class Solution {
  public ListNode middleNode(ListNode head) {
    ListNode fast = head;
    ListNode slow = head;
    while (fast != null && fast.next != null) {
      fast = fast.next.next;
      slow = slow.next;
    }
    return slow;
  }
}
